### PR TITLE
Restrict base-class-first error to class declaration bases

### PR DESCRIPTION
**Pull request by RyanCavanaugh**
Fixes all known repros of #10624

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -16820,7 +16820,9 @@ namespace ts {
                     checkTypeAssignableTo(staticType, getTypeWithoutSignatures(staticBaseType), node.name || node,
                         Diagnostics.Class_static_side_0_incorrectly_extends_base_class_static_side_1);
 
-                    if (baseType.symbol.valueDeclaration && !isInAmbientContext(baseType.symbol.valueDeclaration)) {
+                    if (baseType.symbol.valueDeclaration &&
+                        !isInAmbientContext(baseType.symbol.valueDeclaration) &&
+                        baseType.symbol.valueDeclaration.kind === SyntaxKind.ClassDeclaration) {
```
Comment (mhegazy) [inline src/compiler/checker.ts]: what about:
```ts
class C extends makeClass() {}
var makeClass = function() {...}
```

Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Their declarations are hoisted, which means they might be initialized earlier
```ts
makeClass = function() { ... }
class C extends makeClass() {}
var makeClass = function() {...}
```

Comment (mhegazy) [inline src/compiler/checker.ts]: const/let should give you an error already if used before declaration.

Comment (mhegazy) [inline src/compiler/checker.ts]: so look like it is not worth the change then.. nevermind.

**Pull request merged**